_posts/2023-08-25-using-kani-to-validate-security-boundaries-in-aws-firecracker.md

```diff
@@ -0,0 +1,292 @@
+            LAST_NANOS = next_nanos;
+        }
+
+        0
+    }
+}
+```
+
+Note how the first invocation of this stub will always set `tv_sec = tv_nsec = 0`, as this is what the statics are initialized to. This is an optimization we do because the rate-limiter only cares about the delta between two instants, which will be non-deterministic as long as one of the two instants is non-deterministic. In order to keep Kani performant, it is important to minimize the number of non-deterministic values, especially if multiplication and division is involved.
+
+Using this stub, we can start writing harnesses for `TokenBucket` such as
+
+```rs
+#[kani::proof]
+#[kani::unwind(1)] // enough to unwind the recursion at `Timespec::sub_timespec`
+#[kani::stub(libc::clock_gettime, stubs::clock_gettime)]
+fn verify_token_bucket_auto_replenish() {
+    let mut bucket: TokenBucket = kani::any();
+
+    bucket.auto_replenish();
+
+    // is_valid() performs sanity checks such as "budget <= size".
+    assert!(bucket.is_valid());
+}
+```
+
+Let us now see how we can extend this harness to allow us to verify that our rate limiter is replenishing tokens at exactly the requested rate.
+
+### Verifying the Correctness of our Mitigation
+
+Our noisy neighbor mitigation is correct if we always generate the “correct” number of tokens with each call to `auto_replenish`, meaning it is impossible for a guest to do more I/O than configured. Formally, this means
+
+`0 ≤ (now - last_update) - new_tokens ⋅ (refill_time/size) < refill_time/size`
+
+Here, *new_tokens* is the number of tokens that `auto_replenish` generated. The fraction `refill_time/size` is simply the time it takes to generate a single token. Thus, the property states that if we compute the time that it should have taken to generate *new_tokens* and subtract it from the time that actually passed, we are left with an amount of time less than what it would take to generate an additional token: we replenished the maximal number of tokens possible. 
+
+The difficulty of implementing a correct rate limiter is dealing with “leftover” time: If enough time passed to generate “1.8 tokens”, what does Firecracker do with the “0.8” tokens it cannot (as everything is integer valued) add to the budget? Originally, the rate limiter simply dropped these: if you called `auto_replenish` at an inopportune time, then the “0.8” would not be carried forward and the guest essentially “lost” part of its I/O allowance to rounding. Then, with [#3370](https://github.com/firecracker-microvm/firecracker/pull/3370), we decided to fix this by only advancing `last_update` by `new_tokens ⋅ (refill_time/size)` instead of setting it to *now*. This way the fractional tokens will be carried forward, and we even hand-wrote a [proof](https://github.com/firecracker-microvm/firecracker/pull/3370#pullrequestreview-1252110534) to check that `last_update` and the actual system time will not diverge, boldly concluding
```

Comment:
Just curiosity, can you turn this hand-written proof into a Kani harness?

Reply:
That's pretty much what I tried to do, I tried to take the conclusion of the hand-written proof and turned it into an assertion of the function!